email_handler.py

```diff
@@ -1132,8 +1132,38 @@ def handle_reply(envelope, msg: Message, rcpt_to: str) -> (bool, str):
     # as this is usually included when replying
     if user.replace_reverse_alias:
         LOG.d("Replace reverse-alias %s by contact email %s", reply_email, contact)
+
         msg = replace(msg, reply_email, contact.website_email)
 
+        if config.ENABLE_ALL_REVERSE_ALIAS_REPLACEMENT:
+            start = time.time()
+            last_contact_id = 0
+            # MAX_NB_REVERSE_ALIAS_REPLACEMENT is there to limit potential attack
+            for i in range(0, config.MAX_NB_REVERSE_ALIAS_REPLACEMENT, 100):
```

Comment:
Oh I thought by batch replacement, you meant something like `bulk_replace(msg, reverse_aliases, sender_addresses)` where `reverse_aliases` is a list of reverse alias and `sender_addresses` is list of sender address. This changes all occurrences of reverse aliases by sender instead of iterating through reverse alias 1 by 1. When an email body is complex, the current `replace(msg, old, new)` would take a lot of time as it does a tree traversal through the message body. 

Using `range(0, config.MAX_NB_REVERSE_ALIAS_REPLACEMENT, 100)` is actually going to increase the processing time because of database overhead. I've run a test with config.MAX_NB_REVERSE_ALIAS_REPLACEMENT set to 50000, this takes 25 seconds compared to 0.4s  where we use `.limit(config.MAX_NB_REVERSE_ALIAS_REPLACEMENT)` directly on an alias that has 25 contacts. I recommend to not fetch contacts by batch but rather than fetching the whole Contact object, only fetch what's needed which is contact reply_email and website_email.

Reply:
With the `range(0,config.MAX_...,100)` it certainly takes 4s, but with 1k it takes 0.4s in the production hosts. With the batch size set to that it allows more breathing room to the db.

I haven't found how to do `bulk_replace` in python. But we could try to do something even if it's looping the replace itself, the payload encoding will only be done once.

Reply:
I've reverted the bulk change and added that it only requests the fields it's using